plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/GcpProjectQueryJob.java

```diff
@@ -0,0 +1,90 @@
+import com.google.cloud.tools.eclipse.projectselector.ProjectRepositoryException;
+import com.google.cloud.tools.eclipse.projectselector.ProjectSelector;
+import com.google.cloud.tools.eclipse.projectselector.model.GcpProject;
+import com.google.cloud.tools.eclipse.util.status.StatusUtil;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Predicate;
+import java.util.List;
+import org.eclipse.core.databinding.DataBindingContext;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.core.runtime.jobs.Job;
+import org.eclipse.swt.widgets.Display;
+
+/**
+ * Generic {@link Job} that queries GCP projects of given {@link Credential} through
+ * {@link ProjectRepository} and updates {@link ProjectSelector} asynchronously.
+ */
+public class GcpProjectQueryJob extends Job {
+
+  private final Credential credential;
+  private final ProjectRepository projectRepository;
+  private final ProjectSelector projectSelector;
+  private final DataBindingContext dataBindingContext;
+  private final Predicate<Job> isLatestQueryJob;
+  private final Display display;
+
+  /**
+   * @param projectRepository {@link ProjectRepository#getProjects} must be thread-safe
+   * @param dataBindingContext data binding context that binds {@link #projectSelector}
+   * @param isLatestQueryJob predicate that lazily determines if this job is the latest query job,
+   *     which determines if the job should update {@link ProjectSelector} or die silently. This
+   *     predicate is executed in the UI context
+   */
+  GcpProjectQueryJob(Credential credential, ProjectRepository projectRepository,
+      ProjectSelector projectSelector, DataBindingContext dataBindingContext,
+      Predicate<Job> isLatestQueryJob, Display display) {
+    super("Google Cloud Platform Projects Query Job");
+    this.credential = Preconditions.checkNotNull(credential);
+    this.projectRepository = Preconditions.checkNotNull(projectRepository);
+    this.projectSelector = Preconditions.checkNotNull(projectSelector);
+    this.dataBindingContext = Preconditions.checkNotNull(dataBindingContext);
+    this.isLatestQueryJob = Preconditions.checkNotNull(isLatestQueryJob);
+    this.display = Preconditions.checkNotNull(display);
+  }
+
+  @Override
+  protected IStatus run(IProgressMonitor monitor) {
+    try {
+      final Job thisJob = this;
+      final List<GcpProject> projects = projectRepository.getProjects(credential);
+
+      // The selector may have been disposed (i.e., dialog closed); check it in the UI thread.
+      display.syncExec(new Runnable() {
+        @Override
+        public void run() {
+          if (!projectSelector.isDisposed()
+              && isLatestQueryJob.apply(thisJob) /* intentionally checking in UI context */) {
+            projectSelector.setProjects(projects);
+            dataBindingContext.updateTargets();  // Select saved choice, if any.
```

Comment:
Is this necessary, because `projectSelector.setProjects()` only makes the selection but doesn't update the UI?